File: server_implementation/src/main/java/com/example/localserver/UserService/EitiLabUserService.java
package com.example.localserver.UserService;

import com.example.localserver.DTO.MessageDTO;
import com.example.localserver.DTO.RecordedEventDTO;
import com.example.localserver.DTO.RegisteredLabUserDTO;
import com.example.localserver.Entity.RecordedEvent;
import com.example.localserver.Entity.RegisteredUser;
import com.example.localserver.Repositories.EitiLabUserRepository;
import com.example.localserver.Repositories.RecordedEventRepository;
import com.example.localserver.Util.CryptoUtil;
import com.google.gson.Gson;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Set;

@Service
public class EitiLabUserService {

    private final EitiLabUserRepository eitiLabUserRepository;
    private final RecordedEventRepository recordedEventRepository;

    @Autowired
    public EitiLabUserService(EitiLabUserRepository eitiLabUserRepository, RecordedEventRepository recordedEventRepository) {
        this.eitiLabUserRepository = eitiLabUserRepository;
        this.recordedEventRepository = recordedEventRepository;
    }

    public List<RecordedEventDTO> findStudentRegistries(String email){
        List<RecordedEventDTO> recordedLogsList = new ArrayList<>();
        RegisteredUser searchedUser = eitiLabUserRepository.findStudentByEmail(email);
        if(email != null && searchedUser != null){
            if(email.trim().length() == 18){
                Set<RecordedEvent> recordedEvents = searchedUser.getEventRegistries();
                recordedEvents.stream()
                        .sorted(Comparator.comparing(RecordedEvent::getDateTime))
                        .forEach(el -> recordedLogsList.add(new RecordedEventDTO(email,el.getRegistryContent())));
            }
        }
        return recordedLogsList;
    }

    public ResponseEntity<String> processUserInitData(MessageDTO encryptedMessage) {
        RegisteredLabUserDTO registeredUser =
                (RegisteredLabUserDTO) decryptMessage(encryptedMessage.getValue(),true);
        if (registeredUser != null) {
            if (validateUserData(registeredUser)) {
                RegisteredUser userToSave = new RegisteredUser(registeredUser.getMail(), registeredUser.getUniqueCode());
                eitiLabUserRepository.save(userToSave);
                recordedEventRepository.save(new RecordedEvent("Maszyna zostala zarejestrowana", LocalDateTime.now(), userToSave));
                return ResponseEntity.status(HttpStatus.OK).body("");
            }
            return ResponseEntity.status(HttpStatus.NOT_ACCEPTABLE).body("");
        }
        return ResponseEntity.status(HttpStatus.NOT_ACCEPTABLE).body("");
    }

    public void processEventContent(MessageDTO encryptedMessage) {
        RecordedEventDTO recordedEvent =
                (RecordedEventDTO) decryptMessage(encryptedMessage.getValue(),false);
        RegisteredUser searchedUser=null;
        if (recordedEvent != null && !recordedEvent.getMail().isEmpty()) {
            searchedUser = eitiLabUserRepository.findStudentByEmail(recordedEvent.getMail());
            if(searchedUser != null)
                recordedEventRepository.save(new RecordedEvent(
                        recordedEvent.getRegistryContent(), LocalDateTime.now(), searchedUser));
        }
    }

    private String encryptMessage(String str) {
        String encryptedResponse = "";
        str = fillStringWithChars(str);
        try {
            encryptedResponse = CryptoUtil.encrypt(str);
        } catch (InvalidAlgorithmParameterException | NoSuchPaddingException | IllegalBlockSizeException
                | NoSuchAlgorithmException | BadPaddingException | InvalidKeyException e) {
            e.printStackTrace();
        }
        return encryptedResponse;
    }

    private Object decryptMessage(String encryptedMessage,boolean isRegistration) {
        try {
            String decryptedMessage = CryptoUtil.decrypt(encryptedMessage);
            decryptedMessage = removeCharsFromString(decryptedMessage);
            if(isRegistration)
                return new Gson().fromJson(decryptedMessage, RegisteredLabUserDTO.class);
            else
                return new Gson().fromJson(decryptedMessage, RecordedEventDTO.class);
        } catch (InvalidAlgorithmParameterException | NoSuchPaddingException | IllegalBlockSizeException
                | NoSuchAlgorithmException | BadPaddingException | InvalidKeyException e) {
            e.printStackTrace();
        }
        return null;
    }

    private boolean validateUserData(RegisteredLabUserDTO registeredLabUserDTO) {
        return registeredLabUserDTO.getMail().matches("[0-9]{8}@pw.edu.pl") &&
                registeredLabUserDTO.getUniqueCode().length() == 8;
    }

    private String fillStringWithChars(String str) {
        if (str.length() == 0) {
            return "";
        }
        long stringNecessarySize = str.length();
        stringNecessarySize = (stringNecessarySize % 16 != 0) ? findMaxRangeOfStringLength(str.length(), 0, 16) : stringNecessarySize;
        long numberOfNeededChars = stringNecessarySize - str.length();
        StringBuilder neededChars = new StringBuilder(CryptoUtil.convert_ASCII_To16System((int)numberOfNeededChars));
        for (long i = 0; i < numberOfNeededChars-1; i++) {
            int randomCharIndex = (int)Math.floor(Math.random()*88);
            neededChars.append(CryptoUtil.getWritable_chars().charAt(randomCharIndex));
        }
        str = numberOfNeededChars!=0 ? (neededChars + str) : str;
        return str;
    }

    private long findMaxRangeOfStringLength(long strLength, long lowRange, long highRange) {
        if (strLength >= lowRange && strLength < highRange) {
            return highRange;
        }
        lowRange += 16;
        highRange += 16;
        return findMaxRangeOfStringLength(strLength, lowRange, highRange);
    }

    private String removeCharsFromString(String str) {
        if(str.charAt(0) != '{' && str.charAt(0) != '['){
            String firstASCIIChar = str.substring(0, 1);
            int decimalValue = CryptoUtil.convert_ASCII_To10System(firstASCIIChar);
            str = str.substring(decimalValue);
        }
        return str;
    }
}

File: Server/src/ServerAPI/Services/EitiLabUserService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using ServerAPI.DTOs;
using ServerAPI.Entities;
using ServerAPI.Exceptions;
using ServerAPI.Utility;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace ServerAPI.Repositories
{
	public class EitiLabUserService : IEitiLabUserRepo
	{
		private readonly ServerDBContext _context;
		private readonly ILogger<EitiLabUserService> _logger;

		public EitiLabUserService(ServerDBContext context, ILogger<EitiLabUserService> logger)
		{
			_context = context;
			_logger = logger;
		}

		public bool ProcessUserInitData(MessageDTO encryptedMessage)
		{
			RegisteredLabUserDTO registeredUser = (RegisteredLabUserDTO)DecryptMessage(encryptedMessage.Value, true);

			if (registeredUser == null)
			{
				throw new NotAcceptableException("ERROR");
			}
			var userAfterValidation = ValidateUserData(registeredUser);
			if (userAfterValidation != null && registeredUser != null)
			{
				if (_context.RecordedEventItems
					.Any(u => u.RegisteredUserId == userAfterValidation.Id
					&& u.RegistryContent == "Maszyna zostala zarejestrowana"))
				{
					_context.RecordedEventItems.Add(new RecordedEvent("Ponowne logowanie na maszyne", DateTime.Now, userAfterValidation));
					_context.SaveChanges();
				}
				else
				{
					_context.RecordedEventItems.Add(new RecordedEvent("Maszyna zostala zarejestrowana", DateTime.Now, userAfterValidation));
					_context.SaveChanges();
				}
				return true;
			}
			return false;
		}

		public bool ProcessEventContent(MessageDTO encryptedMessage)
		{
			// JA TU POTRZEBUJE TEZ UNIQUE CODE lub labkę. Bez tego nie rozoznie ktory student dodaje logi.
			RecordedEventDTO recordedEvent = (RecordedEventDTO)DecryptMessage(encryptedMessage.Value, false);
			if (recordedEvent != null && !string.IsNullOrEmpty(recordedEvent.Email))
			{	
				var optionalRequirement = GetRequirement(recordedEvent);
				RegisteredUser searchedUser = FindStudentByEmail(recordedEvent.Email);
				if (searchedUser != null)
				{ 
					_context.RecordedEventItems.Add(new RecordedEvent(recordedEvent.RegistryContent, DateTime.Now, searchedUser));
					if(optionalRequirement!=null)
					{
						AddCompletedRequirement(searchedUser,optionalRequirement);
					}
					_context.SaveChanges();
					return true;
				}
				return false;
			}
			return false;
		}

		private void AddCompletedRequirement(RegisteredUser searchedUser, string optionalRequirement)
		{
			//TBC
			searchedUser.RequirementsCompleted.Add(new LaboratoryRequirement
			{
				ExpirationDate = DateTime.UtcNow,
				Content = optionalRequirement,
				Laboratory = searchedUser.Laboratory,
			});
			_context.RegisteredUserItems.Update(searchedUser);
			_context.SaveChanges();
		}

		private string GetRequirement(RecordedEventDTO recordedEvent)
		{
			if (recordedEvent.RegistryContent.Substring(0, 13) == "bash_command>")
				return recordedEvent.RegistryContent.Substring(13);
			return null;
		}

		private object DecryptMessage(string encryptedMessage, bool isRegistartion)
		{
			try
			{
				string decryptedMessage = Cryptography.Decrypt(encryptedMessage);
				decryptedMessage = RemoveCharsFromString(decryptedMessage);
				if (isRegistartion)
				{
					return Newtonsoft.Json.JsonConvert.DeserializeObject<RegisteredLabUserDTO>(decryptedMessage);
				}
				else
				{
					return Newtonsoft.Json.JsonConvert.DeserializeObject<RecordedEventDTO>(decryptedMessage);
				}
			}
			catch (Exception e)
			{
				Console.WriteLine(e.ToString());
				Console.WriteLine(e.StackTrace);
			}
			return null;
		}

		private string RemoveCharsFromString(string str)
		{
			if (str[0] != '{' && str[0] != '[')
			{
				string firstASCIIChar = str.Substring(0, 1);
				int decimalValue = Cryptography.ConvertASCIITo10System(firstASCIIChar);
				str = str.Substring(decimalValue);
			}
			return str;
		}

		private long FindMaxRangeOfStringLength(long strLength, int lowRange, int highRange)
		{
			if (strLength >= lowRange && strLength < highRange)
			{
				return highRange;
			}
			lowRange += 16;
			highRange += 16;
			return FindMaxRangeOfStringLength(strLength, lowRange, highRange);
		}

		private RegisteredUser ValidateUserData(RegisteredLabUserDTO registeredUser)
		{
			Regex regex = new("[0-9]{8}");
			var check = regex.IsMatch(registeredUser.Email) && registeredUser.UniqueCode.Length == 8;

			if (check)
			{
				var userAuthenticated = _context
				.RegisteredUserItems
				.Include(u => u.EventRegistries)
				.Include(l => l.Laboratory)
				.Include(r => r.Laboratory.LaboratoryRequirements)
				.SingleOrDefault(u => u.Email == registeredUser.Email && u.UniqueCode == registeredUser.UniqueCode);
				return userAuthenticated;
			}
			return null;
		}
		private string EncryptMessage(string str)
		{
			string encryptedResponse = "";
			str = FillStringWithChars(str);
			try
			{
				encryptedResponse = Cryptography.Encrypt(str);
			}
			catch (Exception e)
			{
				Console.WriteLine(e.ToString());
				Console.WriteLine(e.StackTrace);
			}
			return encryptedResponse;
		}

		private string FillStringWithChars(string str)
		{
			if (str.Length == 0)
				return "";

			long stringNecessarySize = str.Length;

			if (stringNecessarySize % 16 != 0)
				stringNecessarySize = FindMaxRangeOfStringLength(str.Length, 0, 16);

			long numberOfNeededChars = stringNecessarySize - str.Length;

			StringBuilder neededChars = new StringBuilder(Cryptography.ConvertASCIITo16System((int)numberOfNeededChars));
			Random random = new Random();

			for (long i = 0; i < numberOfNeededChars - 1; i++)
			{
				int randomCharEmail = (int)Math.Floor(random.NextDouble() * 88);
				neededChars.Append(Cryptography.GetWritableChars()[randomCharEmail]);
			}

			if (numberOfNeededChars != 0)
				str = neededChars + str;

			return str;
		}

		public RegisteredUser FindStudentByEmail(string email)
		{
			//_logger.LogError($"User with id: {id}. GET action invoked");
			var user = _context
				.RegisteredUserItems
				.Include(u => u.EventRegistries)
				.Include(l=>l.Laboratory)
				.Include(r=>r.RequirementsCompleted)
				.SingleOrDefault(u => u.Email == email);

			if (user is null)
			{
				throw new NotAcceptableException("No such student with given email address");
			}
			return user;
		}
	}
}